DEPR: Deprecate str.split return_type

sinhrks opened the pull request:
Closes #9847, Closes #9870. Default value is `expand=False` to be compat with`return_type='series'`. May better to change the default to True in future (and show warning about it)?

CC @jreback @jorisvandenbossche @sreejata

sinhrks committed:
DEPR: Deprecate str.split return_type

jreback commented on pandas/core/strings.py:
```diff
@@ -742,10 +735,7 @@ def str_split(arr, pat=None, n=None, return_type='series'):
             regex = re.compile(pat)
             f = lambda x: regex.split(x, maxsplit=n)
-    if return_type == 'frame':
-        res = DataFrame((Series(x) for x in _na_map(f, arr)), index=arr.index)
```
isn't this what ``expand=True`` does (and btw, can easily fix #10081) at the same time

sinhrks replied:
yes, the logic is replaced to ``_wrap_results_expand``. 

I think #10081 should be considered separately, because we can't simply use workaround for all the cases.

```
s = pd.Series([1.1, '2.2'])

# current behavior
s.str.split('.', expand=True)
#      0    1
# 0  NaN  NaN
# 1    2    2

# workaround or numpy.split
pd.read_table(StringIO(s.to_csv(None, index=None)), sep='.')
#    1  1.1
# 0  2    2
```

jreback commented on pandas/core/strings.py:
```diff
@@ -1096,7 +1089,9 @@ def _wrap_result_expand(self, result, expand=False):
             if expand:
                 result = list(result)
-            return Index(result, name=name)
+                return MultiIndex.from_tuples(result, names=name)
```
this partially closes #10008 yes?

sinhrks replied:
Yes. Updated #10008 for current status. The current impl can return ``MultiIndex`` for ``partition (#9773) already, but it doesn't work for ``split`` (thus changed).

mortada commented on doc/source/whatsnew/v0.16.1.txt:
```diff
@@ -221,6 +221,28 @@ enhancements are performed to make string operation easier.
      s[s.index.str.startswith('a')]
 
+
+- ``split`` now takes ``expand`` keyword to specify returning dimensionality. ``return_type`` is deprecated. (:issue:`9847`)
```
"returning dimensionality" should be "expanding dimensionality"

sinhrks replied:
how about "specify whether to expand dimensionality"?

mortada replied:
Sounds good! 

> On May 8, 2015, at 8:59 PM, Sinhrks <notifications@github.com> wrote:
> 
> In doc/source/whatsnew/v0.16.1.txt:
> 
> > @@ -221,6 +221,28 @@ enhancements are performed to make string operation easier.
> >       idx.str.startswith('a')
> >       s[s.index.str.startswith('a')]
> >  
> > +
> > +- ``split`` now takes ``expand`` keyword to specify returning dimensionality. ``return_type`` is deprecated. (:issue:`9847`)
> how about "specify whether to expand dimensionality"?
> 
> —
> Reply to this email directly or view it on GitHub.
>

mortada commented on pandas/core/strings.py:
```diff
@@ -717,17 +717,10 @@ def str_split(arr, pat=None, n=None, return_type='series'):
     """
-    from pandas.core.series import Series
-    from pandas.core.frame import DataFrame
     from pandas.core.index import Index
```
`Index` doesn't seem to be used in this function anymore, no need to import

sinhrks commented:
@jreback, @mortada Thanks, updated.

jorisvandenbossche commented:
Looks good! 

Another thing I encountered while reviewing: there is no good explanation of the `n` kwarg of split in the docstring, and the default differs between `str_split`/docstring (default of None) and the signature of `split` itself (default of -1)

sinhrks commented:
@jorisvandenbossche Explanation of ``n`` is written as notes, thus moved it to main part. Also, modified the docstring to meet the ``str.split`` without changing the default values. Or may better to change defaults as None and -1 has the same meaning?

I also found current arg names are different from standard ``str.split`` (``sep`` and ``maxsplit`` ) ...

jreback commented:
merged via https://github.com/pydata/pandas/commit/8b8984258cbede113d362c18508f85fcef4eaaf5

@sinhrks thank you!

if you have minor doc changes, pls fee free to make another PR.
